rootfs/api/models/build.py

```diff
@@ -78,9 +79,37 @@ def create(self, user, *args, **kwargs):
             raise DeisException(str(e)) from e
 
     def save(self, **kwargs):
-        removed = {}
         previous_release = self.app.release_set.filter(failed=False).latest()
-        if previous_release.build is not None:
+
+        if (
+            settings.DEIS_DEPLOY_REJECT_IF_PROCFILE_MISSING is True and
+            # previous release had a Procfile and the current one does not
+            (
+                previous_release.build is not None and
+                len(previous_release.build.procfile) > 0 and
+                len(self.procfile) == 0
+            )
+        ):
+            # Reject deployment
+            raise Conflict(
+                'Last deployment had a Procfile but is missing in this deploy. '
+                'For a successful deployment provide a Procfile.'
+            )
+
+        # See if processes are permitted to be removed
+        remove_procs = (
+            # If set to True then contents of Procfile does not affect the outcome
+            settings.DEIS_DEPLOY_PROCFILE_MISSING_REMOVE is True or
```

Comment:
@kmala in that case `app.structure` will be correct (see the tests, I verify `structure` has not changed, unless you mean something else) whereas the latest `build.procfile` will empty.

2 options then, ditch "remove if missing" setting or copy procfile info from the previous build when that setting is enabled and the Procfile info is missing. That'd let things work for scale / deploy.